test/metabase/query_processor_test/expressions_test.clj

```diff
@@ -449,6 +450,234 @@
+(doseq [driver [:mongo :oracle :redshift :sqlite :sqlserver :vertica]]
+  (defmethod driver/database-supports? [driver ::expression-literals]
+    [_driver _feature _database]
+    false))
+
+(deftest ^:parallel basic-literal-expression-test
+  (testing "basic literal expressions"
+    (mt/test-drivers (mt/normal-drivers-with-feature ::expression-literals)
+      (is (= [[1 "" "foo" 0 12345 1.234 true false]
+              [2 "" "foo" 0 12345 1.234 true false]]
+             (mt/formatted-rows
+              standard-literal-expression-row-formats-with-id
+              (mt/run-mbql-query orders
+                {:expressions standard-literal-expression-defs
+                 :fields      (into [$id] standard-literal-expression-refs)
+                 :order-by    [[:asc  $id]]
+                 :limit       2})))))))
+
+(deftest ^:parallel filter-literal-expression-with-=-!=-test
+  (doseq [[and-or eq-ne expected] [[:and :=  [standard-literal-expression-values]]
+                                   [:or  :!= []]]]
+    (testing (str "filter literal expressions with " and-or " " eq-ne)
+      (mt/test-drivers (mt/normal-drivers-with-feature ::expression-literals)
+        (is (= expected
+               (mt/formatted-rows
+                standard-literal-expression-row-formats
+                (mt/run-mbql-query orders
+                  {:expressions standard-literal-expression-defs
+                   :fields      standard-literal-expression-refs
+                   :filter      (into [and-or] (map #(vector eq-ne % %) standard-literal-expression-refs))
+                   :limit       1}))))))))
+
+(deftest ^:parallel nested-literal-expression-test
+  (testing "nested literal expression"
+    ;; TODO Fix this test for H2 (QUE-726)
+    (mt/test-drivers (set/difference (mt/normal-drivers-with-feature ::expression-literals :nested-queries)
+                                     #{:h2})
+      (is (= [(into [1] standard-literal-expression-values)]
+             (mt/formatted-rows
+              standard-literal-expression-row-formats-with-id
+              (mt/run-mbql-query venues
+                {:fields       (into [$id] standard-literal-expression-column-refs)
+                 :source-query {:source-table $$venues
+                                :expressions  standard-literal-expression-defs
+                                :fields       (into [$id] standard-literal-expression-refs)}
+                 :order-by     [[:asc $id]]
+                 :limit        1})))))))
+
+(deftest ^:parallel order-by-literal-expression-test
+  (testing "order-by integer literal expression"
+    ;; ORDER BY a non-negative integer literal is broadly supported (maybe even standard?) and means "order by the Nth
+    ;; column in the select list". Nevertheless, it's questionable whether anyone would actually want to do this.
+    (mt/test-drivers (mt/normal-drivers-with-feature ::expression-literals)
```

Comment:
I think this is mis-compiling the query, and we shouldn't allow it.

It's breaking the semantics expected by a Metabase `:order-by`, if it's actually an indirection to order by some other column. A Metabase `:order-by` on some expression means "based on this expression"!

Reply:
Yeah there is a slack discussion here about whether we should allow order by for literal custom columns. 

https://metaboat.slack.com/archives/C0645JP1W81/p1742317450874479

This is a good point that not only is it questionable, it's actually surprising and probably not what the user intended event if they did try it. What do you think the right point to raise an error would be? In the lib whenever you try to add an order-by on a literal expression?

Reply:
Fixed in latest updates. The test now explicitly verifies that ORDER BY of a literal integer expression does not mean "order by the nth column".